src/rgw/rgw_d3n_datacache.h

```diff
@@ -0,0 +1,366 @@
+      o->lru_next->lru_prev = o->lru_prev;
+    else
+      tail = o->lru_prev;
+    if (o->lru_prev)
+      o->lru_prev->lru_next = o->lru_next;
+    else
+      head = o->lru_next;
+    o->lru_next = o->lru_prev = nullptr;
+  }
+};
+
+
+template <class T>
+class D3nRGWDataCache : public T
+{
+
+  D3nDataCache d3n_data_cache;
+
+public:
+  D3nRGWDataCache() {}
+
+  int init_rados() override {
+    int ret;
+    d3n_data_cache.init(T::cct);
+    ret = T::init_rados();
+    if (ret < 0)
+      return ret;
+
+    return 0;
+  }
+
+  int flush_read_list(struct get_obj_data* d);
+  int get_obj_iterate_cb(const rgw_raw_obj& read_obj, off_t obj_ofs,
+                         off_t read_ofs, off_t len, bool is_head_obj,
+                         RGWObjState *astate, void *arg) override;
+};
+
+
+template<typename T>
+int D3nRGWDataCache<T>::flush_read_list(struct get_obj_data* d) {
+
+  d->d3n_datacache_lock.lock();
+  std::list<bufferlist> l;
+  l.swap(d->d3n_read_list);
+  d->d3n_read_list.clear();
+  d->d3n_datacache_lock.unlock();
+
+  int r = 0;
+
+  std::string oid;
+  std::list<bufferlist>::iterator iter;
+  for (iter = l.begin(); iter != l.end(); ++iter) {
+    bufferlist& bl = *iter;
+    oid = d->d3n_get_pending_oid();
+    if(oid.empty()) {
+      lsubdout(g_ceph_context, rgw, 0) << "ERROR: flush_read_list(): d3n_get_pending_oid() returned empty oid" << dendl;
+      r = -ENOENT;
+      break;
+    }
+    if (bl.length() <= 0x400000)
+      d3n_data_cache.put(bl, bl.length(), oid);
+  }
+
+  return r;
+}
+
+template<typename T>
+int D3nRGWDataCache<T>::get_obj_iterate_cb(const rgw_raw_obj& read_obj, off_t obj_ofs,
+                                 off_t read_ofs, off_t len, bool is_head_obj,
+                                 RGWObjState *astate, void *arg) {
+
+  librados::ObjectReadOperation op;
+  struct get_obj_data* d = static_cast<struct get_obj_data*>(arg);
+  string oid, key;
+
+  int r = 0;
+
+  if (is_head_obj) {
+    // only when reading from the head object do we need to do the atomic test
+    r = T::append_atomic_test(astate, op);
+    if (r < 0)
+      return r;
+
+    if (astate && obj_ofs < astate->data.length()) {
+      unsigned chunk_len = std::min((uint64_t)astate->data.length() - obj_ofs, (uint64_t)len);
+
+      r = d->client_cb->handle_data(astate->data, obj_ofs, chunk_len);
+      if (r < 0)
+        return r;
+
+      len -= chunk_len;
+      d->offset += chunk_len;
+      read_ofs += chunk_len;
+      obj_ofs += chunk_len;
+      if (!len)
+        return 0;
+    }
+  }
+
+  lsubdout(g_ceph_context, rgw, 20) << "D3nRGWDataCache::get_obj_iterate_cb oid=" << read_obj.oid << " obj-ofs=" << obj_ofs << " read_ofs=" << read_ofs << " len=" << len << dendl;
+  op.read(read_ofs, len, nullptr, nullptr);
+
+  const uint64_t cost = len;
+  const uint64_t id = obj_ofs; // use logical object offset for sorting replies
+  oid = read_obj.oid;
+
+  d->d3n_add_pending_oid(oid);
+
+  if (d3n_data_cache.get(read_obj.oid)) {
```

Comment:
i think this should be conditional on `!is_head_obj` too. prefetch (`astate->data`) will usually avoid having to read the head object here, but not in all cases

Reply:
`!is_head_obj` check added